doc/source/data/transforming-datasets.rst

```diff
@@ -78,34 +80,338 @@ express your customized business logic in transformations. Here we will focus on
+======================
 
-.. tip::
+The return type of a batch UDF will determine the format of the resulting dataset. This
+allows you to dynamically convert between data formats during batch transformations.
+
+The following details how each **batch** UDF (as used in
+:meth:`ds.map_batches() <ray.data.Dataset.map_batches>`) return type will be interpreted
+by Datasets when constructing its internal blocks:
 
-   Write your UDFs to leverage built-in vectorized operations in ``pandas.DataFrame``
-   or ``pyarrow.Table`` for better performance. For example, suppose you want to compute
-   the sum of a column in ``pandas.DataFrame``, instead of iterating each row of a batch
-   and sum up values of that column, you may want to use ``df_batch["col_foo"].sum()``.
+.. tabbed:: pd.DataFrame
 
+  Tabular dataset containing
+  `Pandas DataFrame <https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html>`__
+  blocks.
+
+  .. literalinclude:: ./doc_code/transforming_datasets.py
+    :language: python
+    :start-after: __writing_pandas_out_udfs_begin__
+    :end-before: __writing_pandas_out_udfs_end__
+
+.. tabbed:: pa.Table
+
+  Tabular dataset containing
+  `pyarrow.Table <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>`__
+  blocks.
+
+  .. literalinclude:: ./doc_code/transforming_datasets.py
+    :language: python
+    :start-after: __writing_arrow_out_udfs_begin__
+    :end-before: __writing_arrow_out_udfs_end__
+
+.. tabbed:: np.ndarray
+
+  Tensor dataset containing
+  `pyarrow.Table <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>`__
+  blocks, representing the tensor using our
+  :ref:`tensor extension type <dataset-tensor-extension-api>` to embed the tensor in
+  these tables under a single ``"__value__"`` column.
+
+  .. literalinclude:: ./doc_code/transforming_datasets.py
+    :language: python
+    :start-after: __writing_numpy_out_udfs_begin__
+    :end-before: __writing_numpy_out_udfs_end__
+
+.. tabbed:: Dict[str, np.ndarray]
+
+  Tabular dataset containing
+  `pyarrow.Table <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>`__
+  blocks, where each key-value pair treats the key as the column name as the value as
+  the column tensor.
+
+  If a column tensor is 1-dimensional, then the native Arrow 1D list
+  type is used; if a column tensor has 2 or more dimensions, then we use our
+  :ref:`tensor extension type <dataset-tensor-extension-api>` to embed these
+  n-dimensional tensors in the Arrow table.
+
+  .. literalinclude:: ./doc_code/transforming_datasets.py
+    :language: python
+    :start-after: __writing_numpy_dict_out_udfs_begin__
+    :end-before: __writing_numpy_dict_out_udfs_end__
+
+.. tabbed:: list
+
+  Simple dataset containing Python list blocks.
+
+  .. literalinclude:: ./doc_code/transforming_datasets.py
+    :language: python
+    :start-after: __writing_simple_out_udfs_begin__
+    :end-before: __writing_simple_out_udfs_end__
+
+Row UDF Output Types
```

Comment:
I think we can drop this to be consistent with the rest of the documentation which downgrade row-oriented transformations.

Reply:
We can emphasize batch transformations, but IMO we cannot leave the row transformation format inference undocumented.

Reply:
Ok, what about other transformations like row-level filter/flat_map()?

Reply:
We only need to show one of the row-level mappers to document how the row UDF output types correspond to the internal dataset format, and unlike `.map()` for map-based ingest, I'm not aware of any use cases where `.filter()` or `.flat_map()` are the best solutions, so I don't think we need to document them here.